pkg/utils/k8sutil.go

```diff
@@ -17,115 +17,128 @@ limitations under the License.
-		//	host, ns, resourceVersion))
-		return httpClient.Get(fmt.Sprintf("https://%s:8443/apis/k8s.io/v1/functions?watch=true&resourceVersion=%s",
-			host, resourceVersion))
+// GetRestClient returns a k8s restclient to the request from inside of cluster
+func GetRestClient() (*rest.RESTClient, error) {
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, err
 	}
-}
 
-func submitResource(host, ns string, httpClient *http.Client, body io.Reader) (*http.Response, error) {
-	if host == "localhost" {
-		return httpClient.Post(fmt.Sprintf("http://%s:8080/apis/k8s.io/v1/namespaces/%s/functions",
-			host, ns), "application/json", body)
-	} else {
-		return httpClient.Post(fmt.Sprintf("%s/apis/k8s.io/v1/namespaces/%s/functions",
-			host, ns), "application/json", body)
+	restClient, err := rest.RESTClientFor(config)
+	if err != nil {
+		return nil, err
 	}
+
+	return restClient, nil
 }
 
-func deleteResource(host, ns, funcName string, httpClient *http.Client) (*http.Response, error) {
-	var (
-		req *http.Request
-		err error
-	)
+// GetTPRClient returns tpr client to the request from inside of cluster
+func GetTPRClient() (*rest.RESTClient, error) {
+	tprconfig, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	configureClient(tprconfig)
 
-	if host == "localhost" {
-		req, err = http.NewRequest("DELETE", fmt.Sprintf("http://%s:8080/apis/k8s.io/v1/namespaces/%s/functions/%s",
-			host, ns, funcName), nil)
-	} else {
-		req, err = http.NewRequest("DELETE", fmt.Sprintf("%s/apis/k8s.io/v1/namespaces/%s/functions/%s",
-			host, ns, funcName), nil)
+	tprclient, err := rest.RESTClientFor(tprconfig)
+	if err != nil {
+		return nil, err
 	}
 
+	return tprclient, nil
+}
+
+// GetTPRClientOutOfCluster returns tpr client to the request from outside of cluster
+func GetTPRClientOutOfCluster() (*rest.RESTClient, error) {
+	tprconfig, err := clientcmd.BuildConfigFromFlags("", os.Getenv("HOME")+"/.kube/config")
+	if err != nil {
+		logrus.Fatalf("Can not get kubernetes config: %v", err)
+	}
+
+	configureClient(tprconfig)
+
+	tprclient, err := rest.RESTClientFor(tprconfig)
 	if err != nil {
 		return nil, err
 	}
-	resq, err := httpClient.Do(req)
-	return resq, err
+
+	return tprclient, nil
 }
 
-func CreateK8sResources(ns, name string, spec *spec.FunctionSpec, client *client.Client) error {
+// WatchResources looking for changes of custom function objects
+func WatchResources(httpClient *http.Client, resourceVersion string) (*http.Response, error) {
+	host, port := os.Getenv("KUBERNETES_SERVICE_HOST"), os.Getenv("KUBERNETES_SERVICE_PORT")
+	if len(host) == 0 || len(port) == 0 {
+		return nil, fmt.Errorf("unable to load in-cluster configuration, KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT must be defined")
+	}
+
+	return httpClient.Get(fmt.Sprintf("https://%s:%s/apis/k8s.io/v1/functions?watch=true&resourceVersion=%s",
+		host, port, resourceVersion))
```

Comment:
This won't work when `host` is an IPv6 literal.  Use `net.JoinHostPort(host, port)` and/or construct a `url.URL` object and return `url.String()`  (the latter will also make sure that resourceVersion is quoted correctly, etc)